File: cpp/2918_min_sum/min_sum.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  long long minSum(vector<int>& nums1, vector<int>& nums2) {
    long long sum1 = 0, sum2 = 0;
    long long zero1 = 0, zero2 = 0;
    for (int i : nums1) {
      sum1 += i;
      if (i == 0) {
        sum1++;
        zero1++;
      }
    }
    for (int i : nums2) {
      sum2 += i;
      if (i == 0) {
        sum2++;
        zero2++;
      }
    }
    if ((!zero1 && sum2 > sum1) || (!zero2 && sum1 > sum2)) {
      return -1;
    }
    return max(sum1, sum2);
  }
};

int main() {
  vector<tuple<vector<int>, vector<int>, long long>> tests{
      {{3, 2, 0, 1, 0}, {6, 5, 0}, 12},
      {{2, 0, 2, 0}, {1, 4}, -1},
  };

  for (auto& [nums1, nums2, ans] : tests) {
    assert(Solution().minSum(nums1, nums2) == ans);
  }
}

File: rust/2918_min_sum/min_sum.rs
struct Solution;

impl Solution {
    pub fn min_sum(nums1: Vec<i32>, nums2: Vec<i32>) -> i64 {
        let mut sum1: i64 = 0;
        let mut sum2: i64 = 0;
        let mut zero1 = 0;
        let mut zero2 = 0;

        for &x in &nums1 {
            sum1 += x as i64;
            if x == 0 {
                sum1 += 1;
                zero1 += 1;
            }
        }

        for &x in &nums2 {
            sum2 += x as i64;
            if x == 0 {
                sum2 += 1;
                zero2 += 1;
            }
        }

        if (zero1 == 0 && sum2 > sum1) || (zero2 == 0 && sum1 > sum2) {
            return -1;
        }

        sum1.max(sum2)
    }
}

fn main() {
    let tests = vec![
        (vec![3, 2, 0, 1, 0], vec![6, 5, 0], 12),
        (vec![2, 0, 2, 0], vec![1, 4], -1),
    ];

    for (nums1, nums2, ans) in tests {
        assert_eq!(Solution::min_sum(nums1, nums2), ans);
    }
}
